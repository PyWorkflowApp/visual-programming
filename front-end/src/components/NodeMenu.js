import React from 'react';
import * as _ from 'lodash';
import { OverlayTrigger, Tooltip } from 'react-bootstrap';
import CustomNodeUpload from "./CustomNodeUpload";


export default function NodeMenu(props) {
    // construct menu from JSON of node types
    return (
        <div className="NodeMenu">
            <h3>Node Menu</h3>
            <div>Drag-and-drop nodes to build a workflow.</div>
            <hr />
            {_.map(props.nodes, (items, section) =>
                <div key={`node-menu-${section}`}>
                    <span className="node-section-title">{section}</span>
                    <ul>
                        { _.map(items, item => {
                            const data = {...item}; // copy so we can mutate
                            const config = data.options;
                            delete data.options;
                            return (
                                <NodeMenuItem key={data.node_key || data.filename}
                                              nodeInfo={data} config={config} />
                            )}
                        )}
                    </ul>
                </div>
            )}
            <CustomNodeUpload onUpload={props.onUpload} />
       </div>
    );
}


/**
 * Format docstring with newlines into tooltip content
 * @param string - node docstring
 * @returns {array} - array of strings and HTML elements
 */
function formatTooltip(string) {
    const split = string.split("\n");
    const out = [];
    split.forEach((line, i) => {
        out.push(line);
        out.push(<br  key={i} />);
    });
    out.pop();
    return out;
}


function NodeMenuItem(props) {
    if (!props.nodeInfo.missing_packages) {
        const tooltip = props.nodeInfo.doc ? formatTooltip(props.nodeInfo.doc) : "This node has no documentation."
        return (
            <OverlayTrigger
                placement="right"
                delay={{ show: 250, hide: 250 }}
                overlay={<NodeTooltip message={tooltip} />}>
                <li className="NodeMenuItem"
                    draggable={true}
                    onDragStart={event => {
                        event.dataTransfer.setData(
                            'storm-diagram-node',
                            JSON.stringify(props));
                    }}
                    style={{color: props.nodeInfo.color}}>
                    {props.nodeInfo.name}
                </li>
            </OverlayTrigger>
        )
    } else {
        let tooltip = "These Python modules could not be imported:\n\n"
            + props.nodeInfo.missing_packages.join("\n");
        tooltip = formatTooltip(tooltip);
        return (
            <OverlayTrigger
                placement="right"
                delay={{ show: 250, hide: 250 }}
                overlay={<NodeTooltip message={tooltip} />}>
                <li className="NodeMenuItem invalid">{props.nodeInfo.filename}</li>
            </OverlayTrigger>
        )
    }
}


// Overlay with props has to use ref forwarding
const NodeTooltip = React.forwardRef((props, ref) => {
    return (
        <Tooltip {...props} ref={ref}>
            <div style={{textAlign: "left"}}>
                {props.message}
            </div>
        </Tooltip>
    )
});
