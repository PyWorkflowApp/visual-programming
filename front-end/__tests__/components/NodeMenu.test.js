import React from 'react'
import { render } from '@testing-library/react'
import NodeMenu from '../../src/components/NodeMenu';

const nodes = {"Visualization": [
  { "missing_packages": ["pandas", "time"], "name": "Graph Node", "node_key": "GraphNode", "node_type": "visualization", "num_in": 1, "num_out": 0, "color": "red", "filename": "graph", "doc": "Displays a pandas DataFrame in a visual graph.\n\n    Raises:\n        NodeException: any error generating Altair Chart.\n    ", "options": {"graph_type": "bar", "mark_options": false, "width": 10, "height": 10, "encode_options": true, "x_axis": "a", "y_axis": "average(b)"}, "option_types": {"graph_type": {"type": "select", "label": "Graph Type", "value": "bar", "docstring": "Graph viz type", "options": ["area", "bar", "line", "point"]}, "mark_options": {"type": "boolean", "label": "Specify mark options", "value": false, "docstring": "Specify mark options"}, "width": {"type": "int", "label": "Mark width", "value": 10, "docstring": "Width of marks"}, "height": {"type": "int", "label": "Mark height", "value": 10, "docstring": "Height of marks"}, "encode_options": {"type": "boolean", "label": "Specify encoding options", "value": true, "docstring": "Specify encoding options"}, "x_axis": {"type": "string", "label": "X-Axis", "value": "a", "docstring": "X-axis values"}, "y_axis": {"type": "string", "label": "Y-Axis", "value": "average(b)", "docstring": "Y-axis values"}}, "download_result": false}], "Flow Control": [{"name": "Integer Input", "node_key": "IntegerNode", "node_type": "flow_control", "num_in": 1, "num_out": 1, "color": "purple", "filename": "integer_input", "doc": "StringNode object\n\n    Allows for Strings to replace 'string' fields in Nodes\n    ", "options": {"default_value": null, "var_name": "my_var"}, "option_types": {"default_value": {"type": "int", "label": "Default Value", "value": null, "docstring": "Value this node will pass as a flow variable"}, "var_name": {"type": "string", "label": "Variable Name", "value": "my_var", "docstring": "Name of the variable to use in another Node"}}, "download_result": false}, {"name": "String Input", "node_key": "StringNode", "node_type": "flow_control", "num_in": 1, "num_out": 1, "color": "purple", "filename": "string_input", "doc": "StringNode object\n\n    Allows for Strings to replace 'string' fields in Nodes\n    ", "options": {"default_value": null, "var_name": "my_var"}, "option_types": {"default_value": {"type": "string", "label": "Default Value", "value": null, "docstring": "Value this node will pass as a flow variable"}, "var_name": {"type": "string", "label": "Variable Name", "value": "my_var", "docstring": "Name of the variable to use in another Node"}}, "download_result": false}], "Manipulation": [{"name": "Joiner", "node_key": "JoinNode", "node_type": "manipulation", "num_in": 2, "num_out": 1, "color": "goldenrod", "filename": "join", "doc": null, "options": {"on": null}, "option_types": {"on": {"type": "string", "label": "Join Column", "value": null, "docstring": "Name of column to join on"}}, "download_result": false}, {"name": "Filter", "node_key": "FilterNode", "node_type": "manipulation", "num_in": 1, "num_out": 1, "color": "goldenrod", "filename": "filter", "doc": null, "options": {"items": null, "like": null, "regex": null, "axis": null}, "option_types": {"items": {"type": "string", "label": "Items", "value": null, "docstring": "Keep labels from axis which are in items"}, "like": {"type": "string", "label": "Like", "value": null, "docstring": "Keep labels from axis for which like in label == True."}, "regex": {"type": "string", "label": "Regex", "value": null, "docstring": "Keep labels from axis for which re.search(regex, label) == True."}, "axis": {"type": "string", "label": "Axis", "value": null, "docstring": "The axis to filter on."}}, "download_result": false}, {"name": "Pivoting", "node_key": "PivotNode", "node_type": "manipulation", "num_in": 1, "num_out": 3, "color": "goldenrod", "filename": "pivot", "doc": null, "options": {"index": null, "values": null, "columns": null, "aggfunc": "mean", "fill_value": null, "margins": false, "dropna": true, "margins_name": "All", "observed": false}, "option_types": {"index": {"type": "string", "label": "Index", "value": null, "docstring": "Column to aggregate (column, grouper, array or list)"}, "values": {"type": "string", "label": "Values", "value": null, "docstring": "Column name to use to populate new frame's values (column, grouper, array or list)"}, "columns": {"type": "string", "label": "Column Name Row", "value": null, "docstring": "Column(s) to use for populating new frame values. (column, grouper, array or list)"}, "aggfunc": {"type": "string", "label": "Aggregation function", "value": "mean", "docstring": "Function used for aggregation (function, list of functions, dict, default numpy.mean)"}, "fill_value": {"type": "string", "label": "Fill value", "value": null, "docstring": "Value to replace missing values with (scalar)"}, "margins": {"type": "boolean", "label": "Margins name", "value": false, "docstring": "Add all rows/columns"}, "dropna": {"type": "boolean", "label": "Drop NaN columns", "value": true, "docstring": "Ignore columns with all NaN entries"}, "margins_name": {"type": "string", "label": "Margins name", "value": "All", "docstring": "Name of the row/column that will contain the totals when margins is True"}, "observed": {"type": "boolean", "label": "Column Name Row", "value": false, "docstring": "Row number with column names (0-indexed) or \"infer\""}}, "download_result": false}], "I/O": [{"name": "Read CSV", "node_key": "ReadCsvNode", "node_type": "io", "num_in": 0, "num_out": 1, "color": "green", "filename": "read_csv", "doc": "ReadCsvNode\n\n    Reads a CSV file into a pandas DataFrame.\n\n    Raises:\n         NodeException: any error reading CSV file, converting\n            to DataFrame.\n    ", "options": {"file": null, "sep": ",", "header": "infer"}, "option_types": {"file": {"type": "file", "label": "File", "value": null, "docstring": "CSV File"}, "sep": {"type": "string", "label": "Delimiter", "value": ",", "docstring": "Column delimiter"}, "header": {"type": "string", "label": "Header Row", "value": "infer", "docstring": "Row number containing column names (0-indexed)"}}, "download_result": false}, {"name": "Write CSV", "node_key": "WriteCsvNode", "node_type": "io", "num_in": 1, "num_out": 0, "color": "green", "filename": "write_csv", "doc": "WriteCsvNode\n\n    Writes the current DataFrame to a CSV file.\n\n    Raises:\n        NodeException: any error writing CSV file, converting\n            from DataFrame.\n    ", "options": {"file": null, "sep": ",", "index": true}, "option_types": {"file": {"type": "string", "label": "Filename", "value": null, "docstring": "CSV file to write"}, "sep": {"type": "string", "label": "Delimiter", "value": ",", "docstring": "Column delimiter"}, "index": {"type": "boolean", "label": "Write Index", "value": true, "docstring": "Write index as column?"}}, "download_result": true}], "Custom Nodes": [{"name": "Table Creator", "node_key": "TableCreatorNode", "node_type": "custom_nodes", "num_in": 0, "num_out": 1, "color": "green", "filename": "table_creator", "doc": "Accepts raw-text CSV input to create data tables.\n\n    Raises:\n         NodeException: any error reading CSV file, converting\n            to DataFrame.\n    ", "options": {"input": "", "sep": ",", "header": "infer"}, "option_types": {"input": {"type": "text", "label": "Input", "value": "", "docstring": "Text input"}, "sep": {"type": "string", "label": "Delimiter", "value": ",", "docstring": "Column delimiter"}, "header": {"type": "string", "label": "Header Row", "value": "infer", "docstring": "Row number containing column names (0-indexed)"}}, "download_result": false}]};

describe('Validate NodeMenu', () => {
  it('Display NodeMenu', () => {
    const getAvailableNodes = jest.fn(() => "");
    const nodeMenu = render(<NodeMenu
      nodes={nodes}
      onUpload={getAvailableNodes}/>);
    expect(nodeMenu).toMatchSnapshot();
  });

  it('Display NodeMenu with missing packages', () => {
    const getAvailableNodes = jest.fn(() => "");
    const nodeInfo = { missing_packages: ["pandas", "time"]}
    const nodeMenu = render(<NodeMenu
      nodes={nodes}
      onUpload={getAvailableNodes}/>);
    expect(nodeMenu).toMatchSnapshot();
  });
});
