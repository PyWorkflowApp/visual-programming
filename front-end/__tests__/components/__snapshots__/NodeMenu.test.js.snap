// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate NodeMenu Display NodeMenu 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="NodeMenu"
      >
        <h3>
          Node Menu
        </h3>
        <div>
          Drag-and-drop nodes to build a workflow.
        </div>
        <hr />
        <div>
          <span
            class="node-section-title"
          >
            Visualization
          </span>
          <ul>
            <li
              class="NodeMenuItem invalid"
            >
              graph
            </li>
          </ul>
        </div>
        <div>
          <span
            class="node-section-title"
          >
            Flow Control
          </span>
          <ul>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: purple;"
            >
              Integer Input
            </li>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: purple;"
            >
              String Input
            </li>
          </ul>
        </div>
        <div>
          <span
            class="node-section-title"
          >
            Manipulation
          </span>
          <ul>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: goldenrod;"
            >
              Joiner
            </li>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: goldenrod;"
            >
              Filter
            </li>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: goldenrod;"
            >
              Pivoting
            </li>
          </ul>
        </div>
        <div>
          <span
            class="node-section-title"
          >
            I/O
          </span>
          <ul>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: green;"
            >
              Read CSV
            </li>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: green;"
            >
              Write CSV
            </li>
          </ul>
        </div>
        <div>
          <span
            class="node-section-title"
          >
            Custom Nodes
          </span>
          <ul>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: green;"
            >
              Table Creator
            </li>
          </ul>
        </div>
        <input
          style="display: none;"
          type="file"
        />
        <button
          class="btn btn-success btn-sm"
          type="button"
        >
          Add Custom Node
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="NodeMenu"
    >
      <h3>
        Node Menu
      </h3>
      <div>
        Drag-and-drop nodes to build a workflow.
      </div>
      <hr />
      <div>
        <span
          class="node-section-title"
        >
          Visualization
        </span>
        <ul>
          <li
            class="NodeMenuItem invalid"
          >
            graph
          </li>
        </ul>
      </div>
      <div>
        <span
          class="node-section-title"
        >
          Flow Control
        </span>
        <ul>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: purple;"
          >
            Integer Input
          </li>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: purple;"
          >
            String Input
          </li>
        </ul>
      </div>
      <div>
        <span
          class="node-section-title"
        >
          Manipulation
        </span>
        <ul>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: goldenrod;"
          >
            Joiner
          </li>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: goldenrod;"
          >
            Filter
          </li>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: goldenrod;"
          >
            Pivoting
          </li>
        </ul>
      </div>
      <div>
        <span
          class="node-section-title"
        >
          I/O
        </span>
        <ul>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: green;"
          >
            Read CSV
          </li>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: green;"
          >
            Write CSV
          </li>
        </ul>
      </div>
      <div>
        <span
          class="node-section-title"
        >
          Custom Nodes
        </span>
        <ul>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: green;"
          >
            Table Creator
          </li>
        </ul>
      </div>
      <input
        style="display: none;"
        type="file"
      />
      <button
        class="btn btn-success btn-sm"
        type="button"
      >
        Add Custom Node
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Validate NodeMenu Display NodeMenu with missing packages 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="NodeMenu"
      >
        <h3>
          Node Menu
        </h3>
        <div>
          Drag-and-drop nodes to build a workflow.
        </div>
        <hr />
        <div>
          <span
            class="node-section-title"
          >
            Visualization
          </span>
          <ul>
            <li
              class="NodeMenuItem invalid"
            >
              graph
            </li>
          </ul>
        </div>
        <div>
          <span
            class="node-section-title"
          >
            Flow Control
          </span>
          <ul>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: purple;"
            >
              Integer Input
            </li>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: purple;"
            >
              String Input
            </li>
          </ul>
        </div>
        <div>
          <span
            class="node-section-title"
          >
            Manipulation
          </span>
          <ul>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: goldenrod;"
            >
              Joiner
            </li>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: goldenrod;"
            >
              Filter
            </li>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: goldenrod;"
            >
              Pivoting
            </li>
          </ul>
        </div>
        <div>
          <span
            class="node-section-title"
          >
            I/O
          </span>
          <ul>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: green;"
            >
              Read CSV
            </li>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: green;"
            >
              Write CSV
            </li>
          </ul>
        </div>
        <div>
          <span
            class="node-section-title"
          >
            Custom Nodes
          </span>
          <ul>
            <li
              class="NodeMenuItem"
              draggable="true"
              style="color: green;"
            >
              Table Creator
            </li>
          </ul>
        </div>
        <input
          style="display: none;"
          type="file"
        />
        <button
          class="btn btn-success btn-sm"
          type="button"
        >
          Add Custom Node
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="NodeMenu"
    >
      <h3>
        Node Menu
      </h3>
      <div>
        Drag-and-drop nodes to build a workflow.
      </div>
      <hr />
      <div>
        <span
          class="node-section-title"
        >
          Visualization
        </span>
        <ul>
          <li
            class="NodeMenuItem invalid"
          >
            graph
          </li>
        </ul>
      </div>
      <div>
        <span
          class="node-section-title"
        >
          Flow Control
        </span>
        <ul>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: purple;"
          >
            Integer Input
          </li>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: purple;"
          >
            String Input
          </li>
        </ul>
      </div>
      <div>
        <span
          class="node-section-title"
        >
          Manipulation
        </span>
        <ul>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: goldenrod;"
          >
            Joiner
          </li>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: goldenrod;"
          >
            Filter
          </li>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: goldenrod;"
          >
            Pivoting
          </li>
        </ul>
      </div>
      <div>
        <span
          class="node-section-title"
        >
          I/O
        </span>
        <ul>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: green;"
          >
            Read CSV
          </li>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: green;"
          >
            Write CSV
          </li>
        </ul>
      </div>
      <div>
        <span
          class="node-section-title"
        >
          Custom Nodes
        </span>
        <ul>
          <li
            class="NodeMenuItem"
            draggable="true"
            style="color: green;"
          >
            Table Creator
          </li>
        </ul>
      </div>
      <input
        style="display: none;"
        type="file"
      />
      <button
        class="btn btn-success btn-sm"
        type="button"
      >
        Add Custom Node
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
